File: app/src/main/java/sfllhkhan95/doodle/core/PaintCanvas.java
package sfllhkhan95.doodle.core;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.util.DisplayMetrics;

import sfllhkhan95.doodle.utils.DoodleDatabase;


/**
 * @author saifkhichi96
 */
public class PaintCanvas extends Canvas {

    private static final int DEFAULT_BG_COLOR = Color.BLACK;
    private int bgColor = DEFAULT_BG_COLOR;

    private Bitmap bgImage;
    private Bitmap bitmap;
    private String projectName = null;

    public static PaintCanvas loadFromBitmap(DisplayMetrics metrics, Bitmap srcBmp) {
        PaintCanvas canvas = new PaintCanvas(metrics);

        float deviceAspect = metrics.widthPixels / (float) metrics.heightPixels;
        float bmpAspect = srcBmp.getWidth() / (float) srcBmp.getHeight();

        if (deviceAspect != bmpAspect) {
            if (srcBmp.getWidth() >= srcBmp.getHeight()) {
                srcBmp = Bitmap.createBitmap(
                        srcBmp,
                        srcBmp.getWidth() / 2 - srcBmp.getHeight() / 2,
                        0,
                        srcBmp.getHeight(),
                        srcBmp.getHeight()
                );
            } else {
                srcBmp = Bitmap.createBitmap(
                        srcBmp,
                        0,
                        srcBmp.getHeight() / 2 - srcBmp.getWidth() / 2,
                        srcBmp.getWidth(),
                        srcBmp.getWidth()
                );
            }

            if (metrics.widthPixels > metrics.heightPixels) {
                int w = metrics.widthPixels;
                int h = (int) (metrics.widthPixels * srcBmp.getHeight() / (float) srcBmp.getWidth());
                canvas.bgImage = Bitmap.createScaledBitmap(srcBmp, w, h, false);
            } else {
                int h = metrics.heightPixels;
                int w = (int) (metrics.heightPixels * srcBmp.getWidth() / (float) srcBmp.getHeight());
                canvas.bgImage = Bitmap.createScaledBitmap(srcBmp, w, h, false);
            }
        } else {
            canvas.bgImage = Bitmap.createScaledBitmap(srcBmp, metrics.widthPixels, metrics.heightPixels, false);
        }

        return canvas;
    }

    public static PaintCanvas loadFromPath(DisplayMetrics metrics, String bmpPath) {
        Bitmap srcBmp = DoodleDatabase.loadDoodle(bmpPath, metrics.widthPixels, metrics.heightPixels);
        PaintCanvas canvas = loadFromBitmap(metrics, srcBmp);
        canvas.setProjectName(bmpPath);

        return canvas;
    }

    public PaintCanvas(DisplayMetrics metrics) {
        int height = metrics.heightPixels;
        int width = metrics.widthPixels;

        bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.RGB_565);
        setBitmap(bitmap);
    }

    public int getColor() {
        return bgColor;
    }

    public void setColor(int bgColor) {
        this.bgColor = bgColor;
    }

    public Bitmap getBitmap() {
        return bitmap;
    }

    public void drawShapes(Shapes shapes) {
        shapes.paint(this);
    }

    public void drawBackground() {
        if (bgImage == null) {
            drawColor(bgColor);
        } else {
            drawBitmap(bgImage, 0, 0, null);
        }
    }

    private void setProjectName(String title) {
        this.projectName = title;
    }

    public void saveProject() {
        if (projectName == null) {
            DoodleDatabase.saveDoodle(bitmap);
        } else {
            DoodleDatabase.saveDoodle(bitmap, projectName);
        }
    }

}

File: app/src/main/java/sfllhkhan95/doodle/models/PaintCanvas.kt
package sfllhkhan95.doodle.models

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.PointF
import android.graphics.Rect
import androidx.core.content.ContextCompat
import sfllhkhan95.doodle.R
import sfllhkhan95.doodle.utils.ProjectUtils


/**
 * @author saifkhichi96
 */
class PaintCanvas(context: Context, metrics: Rect) : Canvas() {

    private var baseBitmap: Bitmap? = null
    private var bitmapPath: String? = null
    val bitmap: Bitmap

    var color: Int = 0

    init {
        color = ContextCompat.getColor(context, DEFAULT_BG_COLOR)

        val height = metrics.height()
        val width = metrics.width()

        bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.RGB_565)
        setBitmap(bitmap)
    }

    fun drawBackground() {
        if (baseBitmap != null) {
            drawBitmap(baseBitmap!!, 0f, 0f, null)
        } else {
            drawColor(color)
        }
    }

    fun drawPaths(tools: Tools) {
        tools.paint(this)
    }

    fun getColor(touchAt: PointF): Int {
        return try {
            bitmap.getPixel(touchAt.x.toInt(), touchAt.y.toInt())
        } catch (ex: IllegalArgumentException) {
            ex.printStackTrace()
            DEFAULT_BG_COLOR
        }

    }

    fun saveProject() {
        bitmapPath?.let { ProjectUtils[ProjectUtils.timestamp2Name(it)] = bitmap }
                ?: ProjectUtils.create(bitmap)
    }

    fun saveProjectCopy() {
        ProjectUtils.create(bitmap)
    }

    companion object {
        private const val DEFAULT_BG_COLOR = R.color.blue_grey_500

        fun createWithBitmap(context: Context, metrics: Rect, source: Bitmap): PaintCanvas {
            var bitmap = source
            val canvas = PaintCanvas(context, metrics)

            val deviceAspect = metrics.width() / metrics.height().toFloat()
            val bmpAspect = bitmap.width / bitmap.height.toFloat()

            if (deviceAspect != bmpAspect) {
                if (deviceAspect > bmpAspect) { // Device is wider, fit width
                    val targetHeight = bitmap.width / deviceAspect
                    bitmap = Bitmap.createBitmap(
                            bitmap,
                            0,
                            ((bitmap.height - targetHeight) / 2).toInt(),
                            bitmap.width,
                            targetHeight.toInt()
                    )
                } else {
                    val targetWidth = bitmap.height * deviceAspect
                    bitmap = Bitmap.createBitmap(
                        bitmap,
                        ((bitmap.width - targetWidth) / 2).toInt(),
                        0,
                        targetWidth.toInt(),
                        bitmap.height
                    )
                }

                if (metrics.width() > metrics.height()) {
                    val w = metrics.width()
                    val h = (metrics.width() * bitmap.height / bitmap.width.toFloat()).toInt()
                    canvas.baseBitmap = Bitmap.createScaledBitmap(bitmap, w, h, false)
                } else {
                    val h = metrics.height()
                    val w = (metrics.height() * bitmap.width / bitmap.height.toFloat()).toInt()
                    canvas.baseBitmap = Bitmap.createScaledBitmap(bitmap, w, h, false)
                }
            } else {
                canvas.baseBitmap = Bitmap.createScaledBitmap(bitmap, metrics.width(), metrics.height(), false)
            }

            return canvas
        }

        fun createWithBitmapPath(context: Context, metrics: Rect, bitmapPath: String): PaintCanvas {
            val srcBmp = ProjectUtils.open(bitmapPath, metrics.width(), metrics.height())
            val canvas = this.createWithBitmap(context, metrics, srcBmp!!)
            canvas.bitmapPath = bitmapPath

            return canvas
        }
    }

}